src/core/atomic.d

```diff
@@ -173,1279 +56,435 @@ version (AsmX86)
+in ( atomicPtrIsProperlyAligned( here ), "Argument `here` is not properly aligned" )
+{
+    return atomicCompareExchangeStrongNoResult( here, ifThis, writeThis );
+}
 
-        Lcpuid:
+/**
+ * Stores 'writeThis' to the memory referenced by 'here' if the value
+ * referenced by 'here' is equal to the value referenced by 'ifThis'.
+ * The prior value referenced by 'here' is written to `ifThis` and
+ * returned to the user.  This operation is both lock-free and atomic.
+ *
+ * Params:
+ *  here      = The address of the destination variable.
+ *  writeThis = The value to store.
+ *  ifThis    = The address of the value to compare, and receives the prior value of `here` as output.
+ *
+ * Returns:
+ *  true if the store occurred, false if not.
+ */
+bool cas(T,V1,V2)( shared(T)* here, V1* ifThis, V2 writeThis ) pure nothrow @nogc @trusted
+    if ( !is(T == class) && !is(T U : U*) &&  __traits( compiles, { *here = writeThis; } ) )
+in ( atomicPtrIsProperlyAligned( here ), "Argument `here` is not properly aligned" )
+{
+    static if ( __traits(isFloating, T) )
+    {
+        static assert ( __traits(isFloating, V1), "Mismatching argument types." );
+        static if ( T.sizeof == 4 )
+            alias IntTy = uint;
+        else static if ( T.sizeof == 8 )
+            alias IntTy = ulong;
+        return atomicCompareExchangeStrong( cast(IntTy*)here, cast(IntTy*)ifThis, *cast(IntTy*)&writeThis );
+    }
+    else
+        return atomicCompareExchangeStrong!( MemoryOrder.seq, MemoryOrder.seq, T )( cast(T*)here, &cast()*ifThis, cast()writeThis );
+}
 
-            // Slow path: We use cpuid to serialize. This is
-            // significantly slower than mfence, but is the
-            // only serialization facility we have available
-            // on older non-SSE2 chips.
-            push EBX;
+/// Ditto
+bool cas(T,V1,V2)( shared(T)* here, shared(V1)* ifThis, shared(V2) writeThis ) pure nothrow @nogc @trusted
+    if ( is(T == class) && __traits( compiles, { *here = writeThis; } ) )
+in ( atomicPtrIsProperlyAligned( here ), "Argument `here` is not properly aligned" )
+{
+    return atomicCompareExchangeStrong( cast(T*)here, cast(V1*)ifThis, cast()writeThis );
+}
 
-            mov EAX, 0;
-            cpuid;
+/// Ditto
+bool cas(T,V1,V2)( shared(T)* here, shared(V1*)* ifThis, shared(V2)* writeThis ) pure nothrow @nogc @trusted
+    if ( is(T U : U*) && __traits( compiles, { *here = writeThis; } ) )
+in ( atomicPtrIsProperlyAligned( here ), "Argument `here` is not properly aligned" )
+{
+    return atomicCompareExchangeStrong!( MemoryOrder.seq, MemoryOrder.seq, T )( cast(T*)here, &cast()*ifThis, writeThis );
+}
 
-            pop EBX;
+/**
+ * Inserts a full load/store memory fence (on platforms that need it). This ensures
+ * that all loads and stores before a call to this function are executed before any
+ * loads and stores after the call.
+ */
+void atomicFence() nothrow @nogc @safe
+{
+    core.internal.atomic.atomicFence();
```

Comment:
```
static import core.internal.atomic;
core.internal.atomic.atomicFence();
```

or do the static import at module level.

Reply:
How did you determine that? I didn't see any mention of atomicFence in the error output :/
Why can't I use an FQN? It's imported as the second line of the file...

Reply:
I looked at the error message and line number ;)

For future reference it was something like "undefined identifier `core.internal` did you mean to static import `core.internal`?"

As to why you can't use the FQN, not sure, I know that it does work with a static import though.

Reply:
I never saw that error message, the messages I saw were "can't find `internal`" or something.

Reply:
Anyway, next hurdle now...